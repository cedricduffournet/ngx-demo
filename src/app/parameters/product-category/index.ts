export * from './parameter-product-category.component';
