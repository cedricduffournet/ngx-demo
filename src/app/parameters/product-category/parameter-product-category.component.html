<app-product-category-list-view></app-product-category-list-view>
