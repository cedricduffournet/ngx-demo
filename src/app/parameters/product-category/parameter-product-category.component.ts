import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-parameter-product-category',
  templateUrl: './parameter-product-category.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ParameterProductCategoryComponent implements OnInit {
  constructor() {}

  ngOnInit() {}
}
