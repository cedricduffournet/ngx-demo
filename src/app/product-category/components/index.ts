export * from './product-category-add/product-category-add.component';
export * from './product-category-update/product-category-update.component';
export * from './product-category-delete/product-category-delete.component';
export * from './product-category-form/product-category-form.component';
export * from './product-category-item/product-category-item.component';
export * from './product-category-items/product-category-items.component';
