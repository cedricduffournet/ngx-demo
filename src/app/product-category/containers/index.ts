export * from './product-category-add-modal/product-category-add-modal.component';
export * from './product-category-update-modal/product-category-update-modal.component';
export * from './product-category-delete-modal/product-category-delete-modal.component';
export * from './product-category-list-view/product-category-list-view.component';
