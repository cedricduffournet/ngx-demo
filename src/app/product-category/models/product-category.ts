import { schema } from 'normalizr';

export interface ProductCategory {
  id: number;
  name: string;
}

export const productCategorySchema = new schema.Entity('productCategories');
