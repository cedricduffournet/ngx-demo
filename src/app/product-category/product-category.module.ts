import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { ProductCategoryEffects } from '@app/product-category/state/effects';
import {
  ProductCategoryAddComponent,
  ProductCategoryUpdateComponent,
  ProductCategoryDeleteComponent,
  ProductCategoryFormComponent,
  ProductCategoryItemComponent,
  ProductCategoryItemsComponent
} from '@app/product-category/components';
import {
  ProductCategoryAddModalComponent,
  ProductCategoryUpdateModalComponent,
  ProductCategoryDeleteModalComponent,
  ProductCategoryListViewComponent
} from '@app/product-category/containers';
import { reducers } from '@app/product-category/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([ProductCategoryEffects]),
    StoreModule.forFeature('productCategories', reducers)
  ],
  declarations: [
    ProductCategoryListViewComponent,
    ProductCategoryAddComponent,
    ProductCategoryUpdateComponent,
    ProductCategoryDeleteComponent,
    ProductCategoryFormComponent,
    ProductCategoryItemComponent,
    ProductCategoryItemsComponent,
    ProductCategoryAddModalComponent,
    ProductCategoryUpdateModalComponent,
    ProductCategoryDeleteModalComponent
  ],
  entryComponents: [
    ProductCategoryUpdateModalComponent,
    ProductCategoryAddModalComponent,
    ProductCategoryDeleteModalComponent
  ],
  exports: [ProductCategoryListViewComponent]
})
export class ProductCategoryModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
