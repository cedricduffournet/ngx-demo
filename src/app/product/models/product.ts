import { schema } from 'normalizr';
import {
  ProductCategory,
  productCategorySchema
} from '@app/product-category/models/product-category';

export interface ProductBase {
  id: number;
  name: string;
  description: string;
  priceAmount: number;
}

export interface Product extends ProductBase {
  categories: number[];
}

export interface ProductFull extends ProductBase {
  categories: ProductCategory[];
}

export const productSchema = new schema.Entity('products', {
  categories: [productCategorySchema]
});

export const getPrice = (price: number) => {
  if (price > 0) {
    return price / 100;
  }
  return 0;
};
