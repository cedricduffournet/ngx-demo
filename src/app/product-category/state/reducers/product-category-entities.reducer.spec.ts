import { reducer } from '@app/product-category/state/reducers/product-category-entities.reducer';
import * as fromProductCategories from '@app/product-category/state/reducers/product-category-entities.reducer';

import {
  ProductCategoryListViewActions,
  ProductCategoryUpdateModalActions,
  ProductCategoryApiActions
} from '@app/product-category/state/actions';
import { ProductCategory } from '@app/product-category/models/product-category';

describe('ProductCategoryEntitiesReducer', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('LOAD_SUCCESS', () => {
    it('should return state with productCategory entities', () => {
      const productCategories = {
        entities: {
          productCategories: {
            1: {
              name: 'Name1'
            },
            2: {
              name: 'Name2'
            }
          }
        },
        result: [1, 2]
      };
      const action = ProductCategoryApiActions.loadProductCategorySuccess({
        productCategories
      });
      const result = reducer(fromProductCategories.INITIAL_STATE, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('UPDATE', () => {
    const id = 1;
    const productCategory = {
      name: 'NameUpdated'
    } as ProductCategory;
    const data = { id, productCategory };

    it('should set updating to true', () => {
      const action = ProductCategoryUpdateModalActions.updateProductCategory({ data });
      const result = reducer(fromProductCategories.INITIAL_STATE, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('UPDATE_SUCCESS', () => {
    it('should update productCategory id 1, upated should be true and updating false', () => {
      const productCategories = {
        1: {
          id: 1,
          name: 'Name 1'
        },
        2: {
          id: 2,
          name: 'Name 2'
        }
      };
      const initialState = {
        ...fromProductCategories.INITIAL_STATE,
        entities: productCategories
      };

      const productCategory = {
        entities: {
          productCategories: {
            1: {
              name: 'Name 1 updated'
            }
          }
        },
        result: 1
      };
      const action = ProductCategoryApiActions.updateProductCategorySuccess({ productCategory });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('UPDATE_FAILURE & SHOW_MODAL_UPDATE', () => {
    const initialState = {
      ...fromProductCategories.INITIAL_STATE,
      updating: true,
      updated: true,
      entities: {
        1: {
          id: 1,
          name: 'Name'
        }
      }
    };

    it('should set updating to false', () => {
      const error = {
        message: 'error'
      };
      const action = ProductCategoryApiActions.updateProductCategoryFailure({ error });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should set updating to false, updated to false when show modal', () => {
      const action = ProductCategoryListViewActions.showUpdateProductCategoryModal();
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('ADD_SUCCESS', () => {
    const productCategories = {
      1: {
        id: 1,
        name: 'Name 1'
      },
      2: {
        id: 2,
        name: 'Name 2'
      }
    };
    const initialState = {
      ...fromProductCategories.INITIAL_STATE,
      entities: productCategories
    };

    const productCategory = {
      entities: {
        productCategories: {
          3: {
            name: 'Name 3'
          }
        }
      },
      result: 3
    };

    it('should add a new productCategory id 3', () => {
      const action = ProductCategoryApiActions.addProductCategorySuccess({ productCategory });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    const productCategories = {
      1: {
        id: 1,
        name: 'Name 1'
      },
      2: {
        id: 2,
        name: 'Name 2'
      }
    };
    const initialState = {
      ...fromProductCategories.INITIAL_STATE,
      entities: productCategories
    };

    const id = 1;

    it('should set selected productCategory (id 1)', () => {
      const action = ProductCategoryListViewActions.selectProductCategory({ id });
      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('getEntities', () => {
      it('should return entities', () => {
        const result = fromProductCategories.getEntities({
          ...fromProductCategories.INITIAL_STATE,
          entities: {
            1: {
              id: 1,
              name: 'Name 1'
            },
            2: {
              id: 2,
              name: 'Name 2'
            }
          }
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getSelectedId', () => {
      it('should return selected id (1)', () => {
        const result = fromProductCategories.getSelectedId({
          ...fromProductCategories.INITIAL_STATE,
          selectedId: 1
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getUpdating', () => {
      it('should return updating true', () => {
        const result = fromProductCategories.getUpdating({
          ...fromProductCategories.INITIAL_STATE,
          updating: true
        });

        expect(result).toMatchSnapshot();
      });
    });

    describe('getUpdated', () => {
      it('should return updated true', () => {
        const result = fromProductCategories.getUpdated({
          ...fromProductCategories.INITIAL_STATE,
          updated: true
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
