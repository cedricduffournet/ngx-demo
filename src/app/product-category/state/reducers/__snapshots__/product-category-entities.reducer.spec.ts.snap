// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCategoryEntitiesReducer ADD_SUCCESS should add a new productCategory id 3 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name 1",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
    "3": Object {
      "name": "Name 3",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer LOAD_PRODUCT should add a new productCategory id 4 when loading product 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name 1",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
    "4": Object {
      "name": "Name 4",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer LOAD_SUCCESS should return state with productCategory entities 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "name": "Name1",
    },
    "2": Object {
      "name": "Name2",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer SELECT should set selected productCategory (id 1) 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name 1",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
  },
  "selectedId": 1,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer Selectors getEntities should return entities 1`] = `
Object {
  "1": Object {
    "id": 1,
    "name": "Name 1",
  },
  "2": Object {
    "id": 2,
    "name": "Name 2",
  },
}
`;

exports[`ProductCategoryEntitiesReducer Selectors getSelectedId should return selected id (1) 1`] = `1`;

exports[`ProductCategoryEntitiesReducer Selectors getUpdated should return updated true 1`] = `true`;

exports[`ProductCategoryEntitiesReducer Selectors getUpdating should return updating true 1`] = `true`;

exports[`ProductCategoryEntitiesReducer UPDATE should set updating to true 1`] = `
Object {
  "entities": Object {},
  "selectedId": null,
  "updated": false,
  "updating": true,
}
`;

exports[`ProductCategoryEntitiesReducer UPDATE_FAILURE & SHOW_MODAL_UPDATE should set updating to false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer UPDATE_FAILURE & SHOW_MODAL_UPDATE should set updating to false, updated to false when show modal 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "Name",
    },
  },
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer UPDATE_SUCCESS should update productCategory id 1, upated should be true and updating false 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "name": "Name 1 updated",
    },
    "2": Object {
      "id": 2,
      "name": "Name 2",
    },
  },
  "selectedId": null,
  "updated": true,
  "updating": false,
}
`;

exports[`ProductCategoryEntitiesReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "selectedId": null,
  "updated": false,
  "updating": false,
}
`;
